_posts/2023-02-22-user-defined-operators.md

```diff
@@ -0,0 +1,484 @@
+How will the feature evolve in the short and long term?
+
+#### Literal Suffixes
+The very next step for UDVTs is the ability to provide custom literals of user-defined value types.
+The ease of using literal values in expressions is a big part of what makes a type feel native.
+Solidity already provides fractional literals and our intention is to make them usable with custom
+fixed-point number implementations.
+To achieve that we are going to allow defining suffix functions that return UDVT by converting
+literals that are already available in the language:
+
+```solidity
+type UFixed16x2 is uint16;
+
+function uf(int16 mantissa, uint8 exponent) pure suffix returns (UFixed16x2) { ... }
+```
+```solidity
+UFixed16x2 circumference = 2 uf * 3.14 uf * radius;
+```
+
+While the application of a suffix will involve a function call, eventually, with the help of
+compile-time evaluation, literal suffixes are going to become as efficient as native literals.
+
+#### More Operators
+Many operators have been excluded from the initial implementation.
+Some of them will very likely be user-definable at some point.
+
+1. `<<`, `>>` and `**` were excluded due to their non-uniform arguments.
+    For built-in types, the right-hand side is always an unsigned type.
+1.  `!` was excluded because we have to decide whether it should return the same type as the argument
+    of just `bool`.
+    We are leaning towards the latter, but it needs more consideration.
+1. `++` and `--` are very convenient, and we want to have them eventually.
+    They just need to be handled in a special way, and there is more than way to do so.
+1. Composite assignment operators, such as `+=`, will likely be provided automatically
+    in the future as long as the corresponding arithmetic/bitwise operator is available.
+    Allowing a direct implementation is currently not impossible anyway, because there are no references
+    to value types in the language.
+1. We are considering adding a C++-like
+    ["spaceship operator"](https://en.wikipedia.org/wiki/Three-way_comparison#High-level_languages),
+    i.e. `<=>`.
+    When defined, the compiler could use it to automatically generate all the other comparison operators.
+
+On the other hand, it is unlikely that we will allow redefining some other operators:
+
+1. `&&` and `||` are not necessary for most UDVTs because user-defined comparison operators
+    must return `bool`, for which the built-in boolean operators are sufficient.
+    The additional complication is their short-circuiting behavior that may or may not be
+    appropriate, depending on the semantics of a particular UDVT.
+1. `=` is already provided by the compiler for all types and it is very unlikely that we will ever allow redefining it.
+
+#### Operators on Reference Types
+While UDVTs were the primary target, not all types can reasonably fit into a single slot.
+The way to solve this problem is to allow operators taking and returning structs.
```

Comment:
Not sure this tracks. Maybe it's easier to say that we're providing a minimal viable implementation for UDVTs first and consider extending it in the future? We'll probably have multi-stack-slot non-reference types and the way to solve the multi-slot thing is probably *not* always memory structs, etc. :-).

Reply:
How about this?

```
#### Operators on Complex Types
While UDVTs were the primary target, in some cases it may also be reasonable to build a custom type
with a more complex underlying representation, such as a struct or a dynamic array.
The main thing holding us back here are upcoming changes to the type system that are a significant part
of our roadmap for the near future.
To avoid having to introduce breaking changes later, we decided to defer support for such operators until
these changes are behind us.
```

Reply:
The whole section about the future of operators removed.